File: matlab/Tutorials/Conical_Horn_Antenna.m
%
% Tutorials / conical horn antenna
%
% Describtion at:
% http://openems.de/index.php/Tutorial:_Conical_Horn_Antenna
%
% Tested with
%  - Matlab 2011a / Octave 4.0
%  - openEMS v0.0.33
%
% (C) 2011-2015 Thorsten Liebig <thorsten.liebig@uni-due.de>

close all
clear
clc

%% setup the simulation
physical_constants;
unit = 1e-3; % all length in mm

% horn radius
horn.radius  = 20;
% horn length in z-direction
horn.length = 50;

horn.feed_length = 50;

horn.thickness = 2;

% horn opening angle
horn.angle = 20*pi/180;

% size of the simulation box
SimBox = [100 100 100]*2;

% frequency range of interest
f_start =  10e9;
f_stop  =  20e9;

% frequency of interest
f0 = 15e9;

%% setup FDTD parameter & excitation function
FDTD = InitFDTD( 'NrTS', 30000, 'EndCriteria', 1e-4 );
FDTD = SetGaussExcite(FDTD,0.5*(f_start+f_stop),0.5*(f_stop-f_start));
BC = {'PML_8' 'PML_8' 'PML_8' 'PML_8' 'PML_8' 'PML_8'}; % boundary conditions
FDTD = SetBoundaryCond( FDTD, BC );

%% setup CSXCAD geometry & mesh
% currently, openEMS cannot automatically generate a mesh
max_res = c0 / (f_stop) / unit / 15; % cell size: lambda/20
CSX = InitCSX();

%create fixed lines for the simulation box, substrate and port
mesh.x = [-SimBox(1)/2 -horn.radius 0 horn.radius SimBox(1)/2];
mesh.x = SmoothMeshLines( mesh.x, max_res, 1.4); % create a smooth mesh between specified fixed mesh lines

mesh.y = mesh.x;

%create fixed lines for the simulation box and given number of lines inside the substrate
mesh.z = [-horn.feed_length 0 SimBox(3) ];
mesh.z = SmoothMeshLines( mesh.z, max_res, 1.4 );

CSX = DefineRectGrid( CSX, unit, mesh );

%% create horn
% horn + waveguide, defined by a rotational polygon
CSX = AddMetal(CSX, 'Conical_Horn');
p(1,1) = horn.radius+horn.thickness;   % x-coord point 1
p(2,1) = -horn.feed_length;     % z-coord point 1
p(1,end+1) = horn.radius+horn.thickness;   % x-coord point 1
p(2,end) = 0;     % z-coord point 1
p(1,end+1) = horn.radius+horn.thickness + sin(horn.angle)*horn.length; % x-coord point 2
p(2,end) = horn.length; % y-coord point 2
p(1,end+1) = horn.radius + sin(horn.angle)*horn.length; % x-coord point 2
p(2,end) = horn.length; % y-coord point 2
p(1,end+1) = horn.radius;  % x-coord point 1
p(2,end) = 0;     % z-coord point 1
p(1,end+1) = horn.radius;   % x-coord point 1
p(2,end) = -horn.feed_length;     % z-coord point 1
CSX = AddRotPoly(CSX,'Conical_Horn',10,'x',p,'z');

% horn aperture
A = pi*((horn.radius + sin(horn.angle)*horn.length)*unit)^2;

%% apply the excitation %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
start=[-horn.radius -horn.radius mesh.z(10) ];
stop =[+horn.radius +horn.radius mesh.z(1)+horn.feed_length/2 ];
[CSX, port] = AddCircWaveGuidePort( CSX, 0, 1, start, stop, horn.radius*unit, 'TE11', 0, 1);

%%
CSX = AddDump(CSX,'Exc_dump');
start=[-horn.radius -horn.radius mesh.z(8)];
stop =[+horn.radius +horn.radius mesh.z(8)];
CSX = AddBox(CSX,'Exc_dump',0,start,stop);

%% nf2ff calc
start = [mesh.x(9) mesh.y(9) mesh.z(9)];
stop  = [mesh.x(end-8) mesh.y(end-8) mesh.z(end-8)];
[CSX nf2ff] = CreateNF2FFBox(CSX, 'nf2ff', start, stop, 'Directions', [1 1 1 1 0 1]);

%% prepare simulation folder
Sim_Path = 'tmp';
Sim_CSX = 'horn_ant.xml';

[status, message, messageid] = rmdir( Sim_Path, 's' ); % clear previous directory
[status, message, messageid] = mkdir( Sim_Path ); % create empty simulation folder

%% write openEMS compatible xml-file
WriteOpenEMS( [Sim_Path '/' Sim_CSX], FDTD, CSX );

%% show the structure
CSXGeomPlot( [Sim_Path '/' Sim_CSX] );

%% run openEMS
RunOpenEMS( Sim_Path, Sim_CSX);

%% postprocessing & do the plots
freq = linspace(f_start,f_stop,201);

port = calcPort(port, Sim_Path, freq);

Zin = port.uf.tot ./ port.if.tot;
s11 = port.uf.ref ./ port.uf.inc;

% plot reflection coefficient S11
figure
plot( freq/1e9, 20*log10(abs(s11)), 'k-', 'Linewidth', 2 );
ylim([-60 0]);
grid on
title( 'reflection coefficient S_{11}' );
xlabel( 'frequency f / GHz' );
ylabel( 'reflection coefficient |S_{11}|' );

drawnow

%% NFFF contour plots %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% calculate the far field at phi=0 degrees and at phi=90 degrees
thetaRange = (0:2:359) - 180;
disp( 'calculating far field at phi=[0 90] deg...' );
nf2ff = CalcNF2FF(nf2ff, Sim_Path, f0, thetaRange*pi/180, [0 90]*pi/180);

Dlog=10*log10(nf2ff.Dmax);
G_a = 4*pi*A/(c0/f0)^2;
e_a = nf2ff.Dmax/G_a;

% display some antenna parameter
disp( ['radiated power: Prad = ' num2str(nf2ff.Prad) ' Watt']);
disp( ['directivity: Dmax = ' num2str(Dlog) ' dBi'] );
disp( ['aperture efficiency: e_a = ' num2str(e_a*100) '%'] );

%%
% normalized directivity
figure
plotFFdB(nf2ff,'xaxis','theta','param',[1 2]);
drawnow
%   D_log = 20*log10(nf2ff.E_norm{1}/max(max(nf2ff.E_norm{1})));
%   D_log = D_log + 10*log10(nf2ff.Dmax);
%   plot( nf2ff.theta, D_log(:,1) ,'k-', nf2ff.theta, D_log(:,2) ,'r-' );

% polar plot
figure
polarFF(nf2ff,'xaxis','theta','param',[1 2],'logscale',[-40 20], 'xtics', 12);
drawnow
%   polar( nf2ff.theta, nf2ff.E_norm{1}(:,1) )

%% calculate 3D pattern
phiRange = sort( unique( [-180:5:-100 -100:2.5:-50 -50:1:50 50:2.5:100 100:5:180] ) );
thetaRange = sort( unique([ 0:1:50 50:2.:100 100:5:180 ]));

disp( 'calculating 3D far field...' );
nf2ff = CalcNF2FF(nf2ff, Sim_Path, f0, thetaRange*pi/180, phiRange*pi/180, 'Verbose',2,'Outfile','nf2ff_3D.h5');

figure
plotFF3D(nf2ff);        % plot liear 3D far field

%%
E_far_normalized = nf2ff.E_norm{1}/max(nf2ff.E_norm{1}(:));
DumpFF2VTK([Sim_Path '/Conical_Horn_Pattern.vtk'],E_far_normalized,thetaRange,phiRange,'scale',1e-3);

File: python/Tutorials/Conical_Horn_Antenna.py
# -*- coding: utf-8 -*-
"""
 Tutorials / Conical Horn Antenna

 A conical horn antenna fed by a circular waveguide, excited in its
 dominant TE11 mode.  Horn and feed are built as a single rotationally
 symmetric body from a cross-sectional polygon swept about the z-axis.

 Tested with
  - python 3.14
  - openEMS v0.37

 (c) 2026 Thorsten Liebig <thorsten.liebig@gmx.de>

"""

### Import Libraries
import os, tempfile
import numpy as np
import matplotlib.pyplot as plt  # pip install matplotlib

from CSXCAD  import ContinuousStructure
from openEMS import openEMS
from openEMS.physical_constants import *

### Simulation Parameters
## Define the length unit and all geometric parameters of the conical horn.
## `horn_radius` is the inner radius of the feeding circular waveguide; its
## value sets the TE11 cut-off frequency and therefore the useful bandwidth
## of the antenna.
Sim_Path = os.path.join(tempfile.gettempdir(), 'Conical_Horn')
print(f'{Sim_Path=}')

post_proc_only = False
unit = 1e-3     # all length in mm

horn_radius      = 20                # feed waveguide inner radius
horn_length      = 50                # horn length in z-direction
horn_feed_length = 50                # length of the circular feed waveguide
horn_thickness   = 2                 # metal wall thickness
horn_angle       = 20*np.pi/180      # horn opening angle

# size of the simulation box
SimBox = np.array([100, 100, 100])*2

# frequency range of interest
f_start = 10e9
f_stop  = 20e9

# frequency of interest
f0 = 15e9

### FDTD Solver and Excitation Setup
## Initialise the FDTD engine and choose a Gaussian pulse excitation that
## covers the full frequency range of interest in a single simulation run.
## PML absorbing boundaries on all six faces prevent reflections from the
## simulation box edges and emulate an open radiating environment.
FDTD = openEMS(NrTS=30000, EndCriteria=1e-4)
FDTD.SetGaussExcite(0.5*(f_start+f_stop), 0.5*(f_stop-f_start))
FDTD.SetBoundaryCond(['PML_8']*6)

### CSXCAD Geometry and Mesh
## Build the Cartesian mesh that covers both the feeding waveguide (negative
## z) and the radiation half-space above the horn aperture. Fixed lines are
## placed at the waveguide wall and simulation-box boundaries; SmoothMeshLines
## fills in the interior at roughly lambda/15 resolution to resolve the fields
## accurately without oversampling the free-space region.
# currently, openEMS cannot automatically generate a mesh
max_res = C0 / f_stop / unit / 15

CSX = ContinuousStructure()
FDTD.SetCSX(CSX)
mesh = CSX.GetGrid()
mesh.SetDeltaUnit(unit)

# create fixed lines for the simulation box and the waveguide
mesh.SetLines('x', [-SimBox[0]/2, -horn_radius, 0, horn_radius, SimBox[0]/2])
mesh.SmoothMeshLines('x', max_res, ratio=1.4)
mesh.SetLines('y', mesh.GetLines('x'))

mesh.SetLines('z', [-horn_feed_length, 0, SimBox[2]])
mesh.SmoothMeshLines('z', max_res, ratio=1.4)

x_lines = mesh.GetLines('x')
y_lines = mesh.GetLines('y')
z_lines = mesh.GetLines('z')

### Conical Horn Geometry
## Construct the metallic horn and its circular waveguide feed as a single
## rotationally-symmetric body. A cross-sectional polygon is defined in the
## (radial, z) plane and rotated 360 degrees about the z-axis using
## AddRotPoly, which approximates the circular profile on the rectangular
## FDTD grid. The aperture area `A` is precomputed here for use in the
## aperture efficiency calculation during post-processing.
#
# AddRotPoly with norm_dir='x' takes the polygon in the y-z plane, so
# points[0] is the radial distance from the z-axis and points[1] is z.
horn = CSX.AddMetal('Conical_Horn')

r_out = horn_radius + horn_thickness
r_ap  = horn_radius + np.sin(horn_angle)*horn_length   # inner aperture radius
p = np.array([
    [r_out, r_out, r_out + np.sin(horn_angle)*horn_length, r_ap,        horn_radius, horn_radius      ],
    [-horn_feed_length, 0, horn_length,                    horn_length, 0,           -horn_feed_length],
])
horn.AddRotPoly(points=p, norm_dir='x', elevation=0, rot_axis='z',
                angle=[0, 2*np.pi], priority=10)

# horn aperture
A = np.pi * (r_ap*unit)**2

### Waveguide Feed Port
## Excite the dominant TE11 mode inside the circular waveguide using
## AddCircWaveGuidePort. The port spans the lower section of the feed
## waveguide and simultaneously injects the excitation and records the
## incident and reflected wave amplitudes needed to compute S11.
start = [-horn_radius, -horn_radius, z_lines[9]]
stop  = [ horn_radius,  horn_radius, z_lines[0] + horn_feed_length/2]
port = FDTD.AddCircWaveGuidePort(0, start, stop, 'z', horn_radius*unit, 'TE11', 0, 1)

### Excitation Field Dump
## Record a 2-D slice of the electric field at a plane within the feed
## waveguide for visual verification that the TE11 mode profile is being
## launched correctly. Inspecting this dump before analysing results helps
## catch port-placement errors early.
Exc_dump = CSX.AddDump('Exc_dump')
Exc_dump.AddBox([-horn_radius, -horn_radius, z_lines[7]],
                [ horn_radius,  horn_radius, z_lines[7]])

### Near-Field to Far-Field Box
## Place a Huygens surface just inside the PML on five faces to record
## near-field data during the FDTD run. The bottom face (-z direction) is
## excluded because the waveguide feed reaches the lower simulation boundary,
## making that face unsuitable for a Huygens surface.
start = [x_lines[8],  y_lines[8],  z_lines[8]]
stop  = [x_lines[-9], y_lines[-9], z_lines[-9]]
nf2ff = FDTD.CreateNF2FFBox('nf2ff', start, stop, directions=[1, 1, 1, 1, 0, 1])

### Preview Geometry
## Open the AppCSXCAD viewer to inspect the mesh and geometry before
## committing to a full simulation run. Confirming that the horn cross-section
## and port placement look correct here can save significant compute time.
if 0:  # debugging only
    CSX_file = os.path.join(Sim_Path, 'horn_ant.xml')
    if not os.path.exists(Sim_Path):
        os.mkdir(Sim_Path)
    CSX.Write2XML(CSX_file)
    from CSXCAD import AppCSXCAD_BIN
    os.system(AppCSXCAD_BIN + ' "{}"'.format(CSX_file))

### Run Simulation
## Launch the openEMS FDTD solver. The engine iterates time steps until the
## stored energy decays below the EndCriteria threshold, ensuring the
## Fourier-transformed port signals are fully converged before post-processing.
if not post_proc_only:
    FDTD.Run(Sim_Path, cleanup=True)

### Post-Processing and S-Parameter Plot
## Transform the recorded port voltages and currents from the time domain to
## the frequency domain and derive the input reflection coefficient S11.
## A large negative S11 across the band confirms good impedance matching
## between the waveguide feed and the radiating aperture.
freq = np.linspace(f_start, f_stop, 201)
port.CalcPort(Sim_Path, freq)

Zin = port.uf_tot / port.if_tot
s11 = port.uf_ref / port.uf_inc

fig, axis = plt.subplots(num="S11", tight_layout=True)
axis.plot(freq/1e9, 20*np.log10(abs(s11)), 'k-', linewidth=2)
axis.grid()
axis.set_xmargin(0)
axis.set_ylim([-60, 0])
axis.set_title('reflection coefficient $S_{11}$')
axis.set_xlabel('frequency f / GHz')
axis.set_ylabel('reflection coefficient $|S_{11}|$ (dB)')

### Far-Field Radiation Patterns
## Invoke the NF2FF transformation at the centre frequency to obtain the
## antenna directivity as a function of elevation angle for two orthogonal
## azimuth cuts (phi = 0 and phi = 90 degrees). The aperture efficiency
## `e_a` relates the achieved directivity to the theoretical maximum for
## a uniformly illuminated aperture of the same physical area.
thetaRange = np.arange(0, 360, 2) - 180
phiRange   = np.array([0, 90])
print('calculating far field at phi=[0 90] deg...')
res = nf2ff.CalcNF2FF(Sim_Path, f0, thetaRange, phiRange)

Dlog = 10*np.log10(res.Dmax[0])
G_a  = 4*np.pi*A/(C0/f0)**2
e_a  = res.Dmax[0]/G_a

# display some antenna parameter
print(f'radiated power: Prad = {res.Prad[0]} Watt')
print(f'directivity: Dmax = {Dlog:.4g} dBi')
print(f'aperture efficiency: e_a = {e_a*100:.4g}%')

### Normalised Directivity Plots
## Display the elevation-angle radiation pattern on a linear dB scale and
## as a polar diagram for both azimuth cuts. The polar plot reveals the
## main-lobe beamwidth and side-lobe levels that characterise the antenna's
## angular selectivity and gain performance.
E_norm = 20*np.log10(res.E_norm[0]/np.max(res.E_norm[0])) + Dlog

fig, axis = plt.subplots(num="Pattern", tight_layout=True)
axis.plot(thetaRange, E_norm[:, 0], 'k-',  linewidth=2, label='xz-plane ($\\varphi=0^\\circ$)')
axis.plot(thetaRange, E_norm[:, 1], 'r--', linewidth=2, label='yz-plane ($\\varphi=90^\\circ$)')
axis.grid()
axis.set_xmargin(0)
axis.set_ylim([-40, 20])
axis.set_title(f'directivity at {f0/1e9:.0f} GHz')
axis.set_xlabel('theta (deg)')
axis.set_ylabel('directivity (dBi)')
axis.legend()

# polar plot
fig = plt.figure(num="Polar", tight_layout=True)
axis = fig.add_subplot(111, polar=True)
axis.plot(thetaRange/180*np.pi, np.clip(E_norm[:, 0], -40, None), 'k-',
          linewidth=2, label='xz-plane ($\\varphi=0^\\circ$)')
axis.plot(thetaRange/180*np.pi, np.clip(E_norm[:, 1], -40, None), 'r--',
          linewidth=2, label='yz-plane ($\\varphi=90^\\circ$)')
axis.set_rlim([-40, 20])
axis.set_theta_zero_location('N')
axis.set_theta_direction(-1)
axis.set_title(f'directivity (dBi) at {f0/1e9:.0f} GHz')
axis.legend(loc='lower right')

### Three-Dimensional Far-Field Pattern
## Compute the full 3-D radiation pattern by sweeping over a dense grid of
## theta and phi angles. Finer angular spacing is used near the main lobe
## and coarser spacing elsewhere to keep the NF2FF computation fast while
## preserving pattern detail where it matters most.
phiRange_3D   = np.unique(np.concatenate((np.arange(-180, -100,  5), np.arange(-100, -50, 2.5),
                                          np.arange( -50,   50,  1), np.arange(  50, 100, 2.5),
                                          np.arange( 100,  181,  5))))
thetaRange_3D = np.unique(np.concatenate((np.arange(0, 50, 1), np.arange(50, 100, 2),
                                          np.arange(100, 181, 5))))

print('calculating 3D far field...')
res_3D = nf2ff.CalcNF2FF(Sim_Path, f0, thetaRange_3D, phiRange_3D,
                         outfile='nf2ff_3D.h5', verbose=2)

E_far_normalized = res_3D.E_norm[0]/np.max(res_3D.E_norm[0])
theta_g, phi_g = np.meshgrid(thetaRange_3D/180*np.pi, phiRange_3D/180*np.pi, indexing='ij')
x = E_far_normalized*np.sin(theta_g)*np.cos(phi_g)
y = E_far_normalized*np.sin(theta_g)*np.sin(phi_g)
z = E_far_normalized*np.cos(theta_g)

fig = plt.figure(num="3D Pattern", tight_layout=True)
axis = fig.add_subplot(111, projection='3d')
axis.plot_surface(x, y, z, cmap='viridis', linewidth=0, antialiased=True,
                  facecolors=plt.cm.viridis(E_far_normalized))
axis.set_title(f'normalized 3D far field at {f0/1e9:.0f} GHz')
axis.set_xlabel('x'); axis.set_ylabel('y'); axis.set_zlabel('z')
axis.set_box_aspect([1, 1, 1])

# show all plots
plt.show()
